src/runtime_src/core/edge/skd/xrt_skd.cpp

```diff
@@ -35,313 +40,370 @@ namespace xrt {
+      return -EINVAL;
     }
 
-    syslog(LOG_INFO,"Finish soft kernel %s init\n",sk_name);
+    msg = std::string("Finish soft kernel %s init\n") + m_sk_name;
+    xrt_core::message::send(severity_level::debug, "SKD", msg);
     return 0;
   }
 
   XCL_DRIVER_DLLESPEC
   void
   skd::run() {
-    int32_t kernel_return = 0;
+    ffi_arg kernel_return = 0;
     int ret = 0;
-    void* ffi_arg_values[num_args];
+    std::vector<void*> ffi_arg_values(m_kernel_args.size());
     // Buffer Objects
-    int boHandles[num_args];
-    void* bos[num_args];
-    uint64_t boSize[num_args];
+    std::vector<size_t> bo_offsets(m_kernel_args.size());
+    std::vector<void*> bos(m_kernel_args.size());
+    std::vector<size_t> bo_size(m_kernel_args.size());
+    std::vector<int> bo_handles(m_kernel_args.size());
     std::vector<int> bo_list;
-
-    while (1) {
-      ret = waitNextCmd();
-
-      if (ret) {
-	/* We are told to exit the soft kernel loop */
-	syslog(LOG_INFO, "Exit soft kernel %s\n", sk_name);
-	break;
+    Clock::time_point start;
+    Clock::time_point end;
+    Clock::time_point cmd_start;
+    Clock::time_point cmd_end;
+
+    while (true) {
+      ret = wait_next_cmd();
+      cmd_start = Clock::now();
+      if(cmd_end < cmd_start) {
+	  std::string msg = std::string("PS Kernel Command interval = ") + std::to_string((std::chrono::duration_cast<ms_t>(cmd_start - cmd_end)).count());
+	  xrt_core::message::send(severity_level::info, "SKD", msg);
       }
 
-      syslog(LOG_INFO, "Got new kernel command!\n");
+      if (ret && (signal==SIGTERM)) {
+	  // We are told to exit the soft kernel loop
+	  std::string msg = std::string("Exit soft kernel %s\n") + m_sk_name;
+	  xrt_core::message::send(severity_level::info, "SKD", msg);
+	  break;
+      }
 
-      /* Reg file indicates the kernel should not be running. */
-      if (!(args_from_host[0] & 0x1))
+      // Reg file indicates the kernel should not be running.
+      if (!(m_args_from_host[0] & 0x1))
 	continue; //AP_START bit is not set; New Cmd is not available
 
       // FFI PS Kernel implementation
       // Map buffers used by kernel
-      for(int i=0;i<num_args;i++) {
-	if((args[i].index == xrt_core::xclbin::kernel_argument::no_index) && (args[i].hosttype.compare("xrtHandles*")==0)) {
-	  ffi_arg_values[i] = &xrtHandle;
-	} else if(args[i].type == xrt_core::xclbin::kernel_argument::argtype::global) {
-	  uint64_t *buf_addr_ptr = (uint64_t*)(&args_from_host[(args[i].offset+PS_KERNEL_REG_OFFSET)/4]);
-	  uint64_t buf_addr = reinterpret_cast<uint64_t>(*buf_addr_ptr);
-	  uint64_t *buf_size_ptr = (uint64_t*)(&args_from_host[(args[i].offset+PS_KERNEL_REG_OFFSET)/4+2]);
-	  uint64_t buf_size = reinterpret_cast<uint64_t>(*buf_size_ptr);
-	  boSize[i] = buf_size;
-
-	  boHandles[i] = xclGetHostBO(devHdl,buf_addr,buf_size);
-	  bos[i] = xclMapBO(devHdl,boHandles[i],true);
-	  ffi_arg_values[i] = &bos[i];
+      for(int i=0;i<m_kernel_args.size();i++) {
```

Comment:
When there are multiple depend arrays (e.g., bo_offsets, bos, m_kernel_args, etc.) it can become difficult to manage.

By chance, have you though of possible alternative solutions that don't have this requirement.

Reply:
Updated the arrays and made into a single vector of structs.